src/ray/core_worker/task_manager.cc

```
@@ -839,6 +881,53 @@ void TaskManager::RecordMetrics() {
   task_counter_.FlushOnChangeCallbacks();
 }
 
+void TaskManager::RecordTaskStatusEvent(const TaskEntry &task_entry,
+                                        rpc::TaskStatus status) {
+  if (!task_event_buffer_) {
+    return;
+  }
+  // Make task event
+  rpc::TaskEvents task_event;
+  task_event.set_task_id(task_entry.spec.TaskId().Binary());
+  task_event.set_attempt_number(task_entry.spec.AttemptNumber());
+  auto state_updates = task_event.mutable_state_updates();
+  switch (status) {
+  case rpc::TaskStatus::PENDING_ARGS_AVAIL: {
+    // Initialize a new TaskInfoEntry
+    auto task_info = MakeTaskInfoEntry(task_entry.spec);
+    task_event.mutable_task_info()->Swap(&task_info);
+    state_updates->set_pending_args_avail_ts(absl::GetCurrentTimeNanos());
+    break;
+  }
+  case rpc::TaskStatus::SUBMITTED_TO_WORKER: {
+    RAY_CHECK(!task_entry.GetNodeId().IsNil())
+        << "Node ID should have been set on the TaskEntry before updating it's status "
+           "to "
+           "SUBMITTED_TO_WORKER.";
+    // Update the node id
+    state_updates->set_node_id(task_entry.GetNodeId().Binary());
+    state_updates->set_submitted_to_worker_ts(absl::GetCurrentTimeNanos());
+    break;
+  }
+  case rpc::TaskStatus::PENDING_NODE_ASSIGNMENT: {
+    state_updates->set_pending_node_assignment_ts(absl::GetCurrentTimeNanos());
+    break;
+  }
+  case rpc::TaskStatus::FINISHED: {
+    state_updates->set_finished_ts(absl::GetCurrentTimeNanos());
+    break;
+  }
+  case rpc::TaskStatus::FAILED: {
+    state_updates->set_failed_ts(absl::GetCurrentTimeNanos());
+    break;
+  }
+  default: {
+    // Do nothing
```

Hmm should we check fail instead to avoid future missing cases?